### PR TITLE
Optimize the calculation algorithm for CloneSet scaling and updating

### LINKED ISSUE #410: [BUG] Increasing the partition in CloneSet has no effect
**What happened**:
In a given CloneSet resource, if decrease the number of partitions by 1, one of the Pod running previous version if killed and the new one is created. However, the process does not go in reverse if I decrease the partition by 1. There is no effect of increasing the partition actually. 

**What you expected to happen**:
It should be possible to scale down new version and slowly increase the previous version when the partition is increased. If this is a feature that isn't supported, I'd expect the request to be rejected by the API.

**How to reproduce it (as minimally and precisely as possible)**:
Create a new cloneset:
```
apiVersion: apps.kruise.io/v1alpha1
kind: CloneSet
metadata:
  labels:
    app: sample
  name: sample
spec:
  replicas: 5
  selector:
    matchLabels:
      app: sample
  template:
    metadata:
      labels:
        app: sample
    spec:
      containers:
      - name: nginx
        image: nginx:1.18.0-alpine
```
Then, update the Spec to a new image version and include partition:

```
   spec:
      containers:
      - name: nginx
        image: nginx:1.18.1-alpine
    updateStrategy:
      partition: 2
```

Then finally, update the partition value to 4:
```
updateStrategy:
  partition: 4
```

One the last update, you will see no changes in the Pods. The Update Strategy is default, `ReCreate` which makes me think it should be stopping two Pods and starting two new Pods.

**Anything else we need to know?**:
While I am running a slightly older than the latest version of Kruise, I don't see anything related to this in changelog. If this is fixed in a newer version, please let me know and I can try it out.

**Environment**:
- Kruise version: v0.5.0
- Kubernetes version (use `kubectl version`):  `v1.16.4`
- OS (e.g: `cat /etc/os-release`): `Debian GNU/Linux 9 (stretch)`
- Kernel (e.g. `uname -a`): `Linux nginx-deployment-8d9f7c4f-78wjn 4.15.0-1077-azure #82-Ubuntu SMP Thu Mar 26 03:25:09 UTC 2020 x86_64 GNU/Linux`
- Install tools: Helm 
- Others:

### LINKED ISSUE #451: [feature request] Selective Pod deletion satisfies the rule of MaxUnavailable and MaxSurge
**What would you like to be added**:
If `maxSurge = 1, maxUnavailable = 0, specified-delete = 1`,  create a new pod first and then delete the specified-delete pod.

If `maxSurge = 0, maxUnavailable = 1, specified-delete = 2`, delete pod  one by one.

**Why is this needed**:
We'd better not break the rule of MaxUavailable and MaxSurge if we can.

### PR DESCRIPTION
**Pull request by FillZpp**
Signed-off-by: FillZpp <FillZpp.pub@gmail.com>

### Ⅰ. Describe what this PR does
Optimize the calculation algorithm for CloneSet scaling and updating.

Code changes:
1. The `scale` and `update` packages have been amalgamated into `sync` package, which includes the main logic of cloneset reconciling.
2. The `calculateDiffsWithExpectation` function is the most important algorithm in cloneset controller. It calculates the pod numbers to scaling and updating for current CloneSet.

Behavior changes:
1. After increasing partition during canary update, the old cloneset controller will do nothing, but now the controller will rollback pods with **updateRevision** to **currentRevision**.
2. When deleting Pods via specified or replicas, the old cloneset controller will delete them immediately, but now the controller will delete Pods according to maxUnavailable and maxSurge.

### Ⅱ. Does this pull request fix one issue?

fixes #410 

### Ⅲ. List the added test cases (unit test/integration test) if any, please explain if no tests are needed.
pkg/controller/cloneset/sync/cloneset_sync_utils_test.go

### Ⅳ. Describe how to verify it
1. Increase partition during canary update
2. Specified delete pods in a CloneSet with maxSurge and maxUnavailable

### CODE REVIEW

File: pkg/controller/cloneset/sync/cloneset_sync_utils.go [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+	useSurge             int
+	useOldRevisionSurge  int
+	// for update
+	updateDiff           int
+	updateMaxUnavailable int
+}
+
+// This is the most important algorithm in cloneset-controller.
+// It calculates the pod numbers to scaling and updating for current CloneSet.
+func calculateDiffsWithExpectation(cs *appsv1alpha1.CloneSet, pods []*v1.Pod, updateRevision string) (res expectationDiffs) {
```
Comment (Fei-Guo) [inline pkg/controller/cloneset/sync/cloneset_sync_utils.go:47]: I suggest to have a separate document to explain the algorithm. I think code comments are not enough to explain the intentions well. When user observed behaviors that are out of their expectations, they need to have a reference to justify the observations.

File: pkg/controller/cloneset/sync/cloneset_sync_utils.go [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+	"k8s.io/klog"
+	"k8s.io/utils/integer"
+)
+
+type expectationDiffs struct {
+	// scaleDiff is the diff number that should scale
+	// '0' means no need to scale
+	// negative number means need to scale out
+	// positive number means need to scale in
+	scaleDiff int
```
Comment (furykerry) [inline pkg/controller/cloneset/sync/cloneset_sync_utils.go:38]: scaleDiff -> scaleIn

scalediff make me very frustrated about the exact direction of scaling, scaleIn can be more clear

File: pkg/controller/cloneset/sync/cloneset_sync_utils.go [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+	useSurge int
+	// useOldRevisionSurge is part of the useSurge number
+	// it indicates the above number of old revision Pods
+	useOldRevisionSurge int
+
+	// updateDiff is the diff number that should update
+	// '0' means no need to update
+	// positive number means need to update more Pods to updateRevision
+	// negative number means need to update more Pods to currentRevision (rollback)
+	updateDiff int
```
Comment (furykerry) [inline pkg/controller/cloneset/sync/cloneset_sync_utils.go:55]: updateDiff -> updateNum

File: pkg/controller/cloneset/sync/cloneset_sync_utils.go [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+
+type expectationDiffs struct {
+	// scaleDiff is the diff number that should scale
+	// '0' means no need to scale
+	// negative number means need to scale out
+	// positive number means need to scale in
+	scaleDiff int
+	// scaleOldRevisionDiff is part of the scaleDiff number
+	// it indicates the scale number of old revision Pods
+	scaleOldRevisionDiff int
```
Comment (furykerry) [inline pkg/controller/cloneset/sync/cloneset_sync_utils.go:41]: scaleOldRevisionDiff -> scaleInOldRevision

File: pkg/controller/cloneset/sync/cloneset_update.go [hunk 5]
```diff
@@ -311,37 +230,66 @@ func (c *realControl) updatePod(cs *appsv1alpha1.CloneSet, coreControl clonesetc
...
+
+	if strategy.PriorityStrategy != nil {
+		waitUpdateIndexes = updatesort.NewPrioritySorter(strategy.PriorityStrategy).Sort(pods, waitUpdateIndexes)
+	}
+	if strategy.ScatterStrategy != nil {
+		waitUpdateIndexes = updatesort.NewScatterSorter(strategy.ScatterStrategy).Sort(pods, waitUpdateIndexes)
+	}
+
+	// PreparingUpdate first
+	sort.Slice(waitUpdateIndexes, func(i, j int) bool {
```
Comment (furykerry) [inline pkg/controller/cloneset/sync/cloneset_update.go:258]: what is the relative weights of different sort criteria: PriorityStrategy , ScatterStrategy and GetPodLifecycleState?
Is it better to use stable sort (slice -> sliceStable) here ?
please describe them in the func comments.

File: pkg/controller/cloneset/sync/cloneset_update.go [hunk 6]
```diff
@@ -311,37 +230,66 @@ func (c *realControl) updatePod(cs *appsv1alpha1.CloneSet, coreControl clonesetc
...
+		preparingUpdateJ := lifecycle.GetPodLifecycleState(pods[waitUpdateIndexes[j]]) == appspub.LifecycleStatePreparingUpdate
+		if preparingUpdateI != preparingUpdateJ {
+			return preparingUpdateI
+		}
+		return false
+	})
+	return waitUpdateIndexes
+}
+
+func limitUpdateIndexes(coreControl clonesetcore.Control, minReadySeconds int32, diffRes expectationDiffs, waitUpdateIndexes []int, pods []*v1.Pod) []int {
```
Comment (furykerry) [inline pkg/controller/cloneset/sync/cloneset_update.go:269]: please describe the limitation criteria and return values

**Review by furykerry — CHANGES_REQUESTED** [review]

**Review by furykerry — APPROVED** [review]

**Pull request merged**